diff --git a/node_modules/react-native-reanimated/src/reanimated2/animation/decay.ts b/node_modules/react-native-reanimated/src/reanimated2/animation/decay.ts
index 9880096..bcd6de9 100644
--- a/node_modules/react-native-reanimated/src/reanimated2/animation/decay.ts
+++ b/node_modules/react-native-reanimated/src/reanimated2/animation/decay.ts
@@ -61,6 +61,7 @@ export function withDecay(
 
     const VELOCITY_EPS = Platform.OS !== 'web' ? 1 : 1 / 20;
     const SLOPE_FACTOR = 0.1;
+    const MIN_DERIVATE_VALUE = 0.1;
 
     let decay: (animation: InnerDecayAnimation, now: number) => boolean;
 
@@ -75,16 +76,21 @@ export function withDecay(
         } = animation;
 
         const deltaTime = Math.min(now - lastTimestamp, 64);
-        const clampIndex = initialVelocity > 0 ? 1 : 0;
         let derivative = 0;
+
+        if (animation.current <= config.clamp![0]) {
+          animation.current = config.clamp![0];
+        } else if (
+          animation.current >= config.clamp![1]
+        ) {
+          animation.current = config.clamp![1];
+        }
+
         if (current < config.clamp![0] || current > config.clamp![1]) {
-          derivative = current - config.clamp![clampIndex];
+          derivative = current - animation.current;
         }
 
-        if (derivative !== 0) {
-          animation.springActive = true;
-        } else if (derivative === 0 && animation.springActive) {
-          animation.current = config.clamp![clampIndex];
+        if (Math.abs(derivative) < MIN_DERIVATE_VALUE) {
           return true;
         }
 
